server/users_controllers.go

```
@@ -0,0 +1,129 @@
+package server
+
+import (
+	"catalyst/internal"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang-jwt/jwt/v5"
+	"golang.org/x/crypto/bcrypt"
+	"gorm.io/gorm"
+)
+
+const (
+	DuplicateUsernameErr = "username is already registered"
+	DuplicateEmailErr    = "email is already registered"
+	BadRequestBodyErr    = "all fields are required"
+	InternalServerErr    = "internal server error"
+	IncorrectPasswordErr = "password is not correct"
+	IncorrectEmailErr    = "email is not correct"
+
+	UserCreatedResponseMessage  = "user is created successfully"
+	UserSignedInResponseMessage = "user signed in successfully"
+)
+
+var signUpRequestBody struct {
+	Username string `binding:"required"`
+	Password string `binding:"required"`
+	Email    string `binding:"required,email"`
+}
+
+func (s *Server) SignUp(ctx *gin.Context) {
+
+	db := s.DBClient
+
+	if err := ctx.Bind(&signUpRequestBody); err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"message": BadRequestBodyErr,
+		})
+		return 
+	}
+
+	if _, err := db.GetUserByEmail(signUpRequestBody.Email); err != gorm.ErrRecordNotFound {
```

other errors are not handled. check the rest too